{"id":619939,"title":"72｜初识异常：异常的产生与分类","content":"<p><strong>本讲相关链接</strong><br>\n异常的分类：<a href=\"https://docs.python.org/zh-cn/3.10/library/exceptions.html#base-classes\">https://docs.python.org/zh-cn/3.10/library/exceptions.html#base-classes</a><br>\n异常的层次结构：<a href=\"https://docs.python.org/zh-cn/3.10/library/exceptions.html#exception-hierarchy\">https://docs.python.org/zh-cn/3.10/library/exceptions.html#exception-hierarchy</a></p><p><strong>课后习题</strong><br>\n请你通过学习过的列表、字典、数字、字符串等数据类型，以及 if、for、while 等条件控制语句，自行编写程序， 引发至少 5 个异常并查看异常类型。</p><p><strong>课程代码、课件及其他相关资料地址</strong><br>\n<a href=\"https://gitee.com/wilsonyin/zero-basics-python\">https://gitee.com/wilsonyin/zero-basics-python</a></p>","comments":[{"had_liked":false,"id":365606,"user_name":"PatrickL","can_delete":false,"product_type":"c3","uid":1341431,"ip_address":"上海","ucode":"EF0EC18639B9B2","user_header":"https://static001.geekbang.org/account/avatar/00/14/77/f7/11548247.jpg","comment_is_top":false,"comment_ctime":1672820749,"is_pvip":false,"replies":null,"discussion_count":0,"race_medal":0,"score":2,"product_id":100310001,"comment_content":"print(1&#47;0)\n# ZeroDivisionError: division by zero\n\nprint(a)\n# NameError: name &#39;a&#39; is not defined\n\nlist1 = [1,2,3]\nprint(list1[3])\n# IndexError: list index out of range\n\ntup = (1,2)\ntup[0] = 3\n# TypeError: &#39;tuple&#39; object does not support item assignment\n\ndict1 = {&#39;a&#39;:1}\nprint(dict1[&#39;b&#39;])\n# KeyError: &#39;b&#39;","like_count":2},{"had_liked":false,"id":385080,"user_name":"Geek_631607","can_delete":false,"product_type":"c3","uid":3789991,"ip_address":"广东","ucode":"EAF874838F0BE4","user_header":"","comment_is_top":false,"comment_ctime":1702022893,"is_pvip":false,"replies":null,"discussion_count":0,"race_medal":0,"score":2,"product_id":100310001,"comment_content":"以下是一些 Python 代码示例，它们会引发至少 5 种不同的异常：\n\n1. **IndexError**：当你试图访问一个序列（如列表或字符串）中不存在的索引时，会引发这个异常。\n\n```python\nmy_list = [1, 2, 3]\nprint(my_list[10])  # IndexError: list index out of range\n```\n\n2. **KeyError**：当你试图访问字典中不存在的键时，会引发这个异常。\n\n```python\nmy_dict = {&#39;a&#39;: 1, &#39;b&#39;: 2, &#39;c&#39;: 3}\nprint(my_dict[&#39;d&#39;])  # KeyError: &#39;d&#39;\n```\n\n3. **ValueError**：当一个函数的参数有正确的类型，但是值不合适，或者你试图将不合适的值转换为特定的类型时，会引发这个异常。\n\n```python\nint(&#39;abc&#39;)  # ValueError: invalid literal for int() with base 10: &#39;abc&#39;\n```\n\n4. **TypeError**：当操作或函数应用于不适当的类型时，会引发这个异常。\n\n```python\n&#39;2&#39; + 2  # TypeError: can only concatenate str (not &quot;int&quot;) to str\n```\n\n5. **ZeroDivisionError**：当你试图除以零时，会引发这个异常。\n\n```python\n1 &#47; 0  # ZeroDivisionError: division by zero\n```\n\n请注意，这些代码只是为了演示如何引发异常，而在实际编程中，我们应该尽量避免这些错误。                                                                                                   .","like_count":0},{"had_liked":false,"id":382648,"user_name":"MarkG","can_delete":false,"product_type":"c3","uid":3675806,"ip_address":"山东","ucode":"454082595FE769","user_header":"https://static001.geekbang.org/account/avatar/00/38/16/9e/48621655.jpg","comment_is_top":false,"comment_ctime":1697725918,"is_pvip":false,"replies":null,"discussion_count":0,"race_medal":0,"score":2,"product_id":100310001,"comment_content":"my_list = [1,2]\nmy_dict = { &quot;a&quot;: 1, &quot;b&quot;: 2 }\nmy_str = &#39;abc&#39;\nmy_tuple = (1,2)\n\n# my_list[2] &#47;&#47; IndexError: list index out of range\n# my_dict[&quot;c&quot;] &#47;&#47; KeyError: &#39;c&#39;\n# print(my_num) &#47;&#47; NameError: name &#39;my_num&#39; is not defined\n# my_tuple.append(3) &#47;&#47; AttributeError: &#39;tuple&#39; object has no attribute &#39;append&#39;\nint(my_str) &#47;&#47; ValueError: invalid literal for int() with base 10: &#39;abc&#39;","like_count":0},{"had_liked":false,"id":369250,"user_name":"yanyu-xin","can_delete":false,"product_type":"c3","uid":1899757,"ip_address":"广东","ucode":"3AA389F9E4C236","user_header":"","comment_is_top":false,"comment_ctime":1677291119,"is_pvip":false,"replies":null,"discussion_count":0,"race_medal":0,"score":2,"product_id":100310001,"comment_content":"a = [1, 3, 4]\nprint(a[5]) # IndexError: list index out of range\nprint(b) # NameError: name &#39;b&#39; is not defined\n\nd= {&#39;a&#39;:2, &#39;c&#39;:4}\nprint(d[&#39;d&#39;]) # KeyError: &#39;d&#39;\n\nc = 123\nprint(c[0]) # TypeError: &#39;int&#39; object is not subscriptable\nprint(c &#47; 0) # ZeroDivisionError: division by zero\n\nd  = &#39;asfdfd&#39;\nprint(d + 3)  # TypeError: can only concatenate str (not &quot;int&quot;) to str\n\nif  i &lt;&gt; 3 : # SyntaxError: invalid syntax\n\nfor i in range(10.3): #TypeError: &#39;float&#39; object cannot be interpreted as an integer\n\nwhile True:\nprint() # IndentationError: expected an indented block","like_count":0}]}